experimental/language_structure/psl/psl_model_dstc_synthetic.py

```diff
@@ -0,0 +1,388 @@
+        previous_statement = self.predicates['previous_statement']
+        return_loss = None
+        for [current_state_index, previous_state_index] in self.state_transitions:
+            previous_state = self._get_tensor_column(logits, previous_state_index)
+            current_state = self._get_tensor_column(logits, current_state_index)
+            has_tfidf = self.predicates[str(current_state_index) + '_sys_index']
+
+            if return_loss is None:
+                return_loss = self.template_rxy_and_sy_and_tx_implies_ux(previous_statement,
+                                                                         previous_state,
+                                                                         has_tfidf,
+                                                                         current_state)
+            else:
+                return_loss += self.template_rxy_and_sy_and_tx_implies_ux(previous_statement,
+                                                                          previous_state,
+                                                                          has_tfidf,
+                                                                          current_state)
+
+        return return_loss
+
+    def generate_predicates(self, data: tf.Tensor):
+        self.predicates['first_statement'] = self._first_statement()
+        self.predicates['last_statement'] = self._last_statement(data)
+        self.predicates['previous_statement'] = self._previous_statement(data)
+
+        for class_name in self.config['words']:
+            usr_index = self.config['words'][class_name]['usr']['index']
+            sys_index = self.config['words'][class_name]['sys']['index']
+            self.predicates[class_name + '_usr_index'] = self._has_word(data, usr_index)
+            self.predicates[class_name + '_sys_index'] = self._has_word(data, sys_index)
+
+    def compute_total_loss(self, data: tf.Tensor, logits: tf.Tensor) -> float:
+        """Calculate the loss for all PSL rules."""
+        total_loss = 0.0
+        rule_kwargs = dict(logits=logits, data=data)
+
+        for rule_weight, rule_function in zip(self.rule_weights, self.rule_functions):
+            total_loss += rule_weight * tf.reduce_sum(rule_function(**rule_kwargs))
+
+        return total_loss
+
+    def compute_loss_per_rule(self, data: tf.Tensor,
+                              logits: tf.Tensor) -> List[float]:
+        """Calculate the loss for each of the PSL rules."""
+        rule_kwargs = dict(logits=logits, data=data)
+        losses = []
+
+        for rule_weight, rule_function in zip(self.rule_weights,
+                                              self.rule_functions):
+            losses.append(rule_weight * rule_function(**rule_kwargs))
+        return losses
+
+    def compute_loss(self, data: tf.Tensor, logits: tf.Tensor) -> float:
+        """Calculate the total loss for all PSL rules."""
+        return sum(self.compute_loss_per_rule(data, logits))
```

Comment:
Is it possible to fold this into `compute_total_loss`?

Reply:
I want to mirror the multiwoz synthetic psl functions, therefore instead of folding this into compute_total_loss, it makes sense to remove compute_total_loss.

Reply:
Thanks for the change!